File: classified/packages/plugin-autocoder/src/actions/generate-code.ts
import {
  Action,
  ActionResult,
  HandlerCallback,
  type IAgentRuntime,
  Memory,
  State,
  elizaLogger,
} from '@elizaos/core';
import { CodeGenerationService } from '../services/CodeGenerationService';
import type { ProjectType } from '../types/index';
import { FormsService, type Form } from '@elizaos/plugin-forms';

export const generateCodeAction: Action = {
  name: 'GENERATE_CODE',
  similes: [
    'BUILD_CODE',
    'CREATE_CODE',
    'WRITE_CODE',
    'DEVELOP_CODE',
    'IMPLEMENT_CODE',
    'CODE_GENERATION',
    'GENERATE_PROJECT',
    'BUILD_PROJECT',
  ],
  description: 'Generate code for ElizaOS projects using AI',

  validate: async (runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    const text = message.content.text?.toLowerCase() || '';

    // Check if this is a form submission
    const data = message.content.data as { action?: string };
    if (data?.action === 'GENERATE_CODE') {
      return true;
    }

    // Check for code generation keywords
    const genKeywords = ['generate', 'create', 'build', 'write', 'implement', 'code'];
    const contextKeywords = ['plugin', 'agent', 'action', 'provider', 'service'];

    const hasGenKeyword = genKeywords.some((keyword) => text.includes(keyword));
    const hasContextKeyword = contextKeywords.some((keyword) => text.includes(keyword));

    return hasGenKeyword && hasContextKeyword;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: any,
    callback?: HandlerCallback,
    _responses?: Memory[]
  ): Promise<ActionResult> => {
    elizaLogger.info('GENERATE_CODE action triggered');

    const codeGenService = runtime.getService('code-generation') as CodeGenerationService;
    if (!codeGenService) {
      elizaLogger.error('Code generation service not available');
      if (callback) {
        await callback({
          text: "I can't generate code right now - the code generation service isn't available. Please ensure all required API keys are configured.",
          error: true,
        });
      }
      return {
        success: false,
        error: 'Code generation service not available',
      };
    }

    const formsService = runtime.getService('forms') as FormsService | null;

    // Extract project type and data
    const projectType = await extractProjectType(runtime, message, state);
    elizaLogger.info(`Project type: ${projectType}`);

    // Handle form flow
    if (formsService) {
      // Check if we have active forms
      const activeForms = await formsService.listForms('active');
      const projectForm = activeForms.find((f: Form) => f.id.startsWith('project-'));

      if (projectForm && projectForm.status !== 'completed') {
        // We have an incomplete form, prompt for next step
        const nextStepPrompt = getNextStepPrompt(projectForm, message.content.text || '');

        if (callback) {
          await callback({
            text: nextStepPrompt,
            form: projectForm,
          });
        }

        return {
          success: true,
          data: {
            status: 'collecting_requirements',
            formId: projectForm.id,
          },
        };
      }

      // Check if this is a form submission
      const messageData = message.content.data as {
        formId?: string;
        formData?: any;
        action?: string;
        projectId?: string;
        projectType?: string;
      };
      if (messageData?.formId && messageData?.formData) {
        // Process form submission
        const formData = messageData.formData;
        const projectData = extractProjectData(formData);

        // Start generation
        if (callback) {
          await callback({
            text: `🚀 Starting code generation for your ${projectData.targetType}...\n\nThis may take a few minutes. I'll:\n1. Research best practices\n2. Generate the code structure\n3. Implement all features\n4. Run quality checks`,
          });
        }

        const result = await codeGenService.generateCode({
          projectName: projectData.projectName,
          description: projectData.description,
          requirements: projectData.requirements,
          apis: projectData.apis || [],
          targetType: projectData.targetType,
          testScenarios: projectData.testScenarios || [],
        });

        if (!result.success) {
          if (callback) {
            await callback({
              text: `❌ Code generation failed: ${result.errors?.join('\n') || 'Unknown error'}`,
              error: true,
            });
          }
          return {
            success: false,
            error: result.errors?.join('\n') || 'Code generation failed',
          };
        }

        // Prepare success response
        let successMessage = '✅ **Code generation complete!**\n\n';
        successMessage += `📁 **Project:** ${projectData.projectName}\n`;
        successMessage += `📋 **Type:** ${projectData.targetType}\n`;

        if (result.files) {
          successMessage += `📄 **Files generated:** ${result.files.length}\n`;
        }

        if (result.projectPath) {
          successMessage += `\n📍 **Project location:** ${result.projectPath}`;
        }

        if (result.executionResults) {
          successMessage += '\n\n**Quality Checks:**\n';
          successMessage += `${result.executionResults.lintPass ? '✅' : '❌'} Linting\n`;
          successMessage += `${result.executionResults.typesPass ? '✅' : '❌'} Type checking\n`;
          successMessage += `${result.executionResults.testsPass ? '✅' : '❌'} Tests\n`;
          successMessage += `${result.executionResults.buildPass ? '✅' : '❌'} Build\n`;
        }

        successMessage += `\n\n🎉 Your ${projectData.targetType} is ready! Let me know if you need any modifications.`;

        if (callback) {
          await callback({
            text: successMessage,
            data: {
              projectPath: result.projectPath,
              files: result.files,
            },
          });
        }

        return {
          success: true,
          data: {
            projectPath: result.projectPath,
            filesGenerated: result.files?.length || 0,
            executionResults: result.executionResults,
          },
        };
      }
    }

    // Direct code generation without form
    const requirements = extractRequirements(message.content.text || '');

    if (!requirements || requirements.length === 0) {
      if (callback) {
        await callback({
          text: 'I need more details about what you want to build. Could you describe:\n- What type of project (plugin, agent, etc.)?\n- What features or functionality it should have?\n- Any specific APIs or services it should use?',
          error: false,
        });
      }
      return {
        success: false,
        error: 'Insufficient requirements provided',
      };
    }

    const projectName = extractProjectName(message.content.text || '') || `project-${Date.now()}`;

    if (callback) {
      await callback({
        text: `🚀 Starting code generation based on your requirements...\n\nGenerating: ${projectName}\nType: ${projectType}`,
      });
    }

    const result = await codeGenService.generateCode({
      projectName,
      description: message.content.text || '',
      requirements,
      apis: extractAPIs(message.content.text || ''),
      targetType: projectType as ProjectType,
    });

    if (!result.success) {
      if (callback) {
        await callback({
          text: `❌ Code generation failed: ${result.errors?.join('\n') || 'Unknown error'}`,
          error: true,
        });
      }
      return {
        success: false,
        error: result.errors?.join('\n') || 'Code generation failed',
      };
    }

    let responseText = '✅ Code generation complete!\n\n';
    responseText += `📁 Project: ${projectName}\n`;

    if (result.files) {
      responseText += `📄 Files generated: ${result.files.length}\n`;
    }

    if (result.projectPath) {
      responseText += `📍 Location: ${result.projectPath}\n`;
    }

    responseText += '\n🎉 Your code is ready!';

    if (callback) {
      await callback({
        text: responseText,
        data: result,
      });
    }

    return {
      success: true,
      data: result,
    };
  },

  examples: [
    [
      {
        name: '{{user1}}',
        content: {
          text: 'Generate a plugin that monitors crypto prices and sends alerts',
        },
      },
      {
        name: '{{agent}}',
        content: {
          text: "I'll generate a crypto price monitoring plugin for you. Let me create the code structure with price tracking and alert functionality.",
          action: 'GENERATE_CODE',
        },
      },
    ],
  ],
};

// Helper functions
async function extractProjectType(
  runtime: IAgentRuntime,
  message: Memory,
  _state?: State
): Promise<string> {
  const text = message.content.text?.toLowerCase() || '';

  if (text.includes('plugin')) {
    return 'plugin';
  }
  if (text.includes('agent') || text.includes('bot')) {
    return 'agent';
  }
  if (text.includes('workflow')) {
    return 'workflow';
  }
  if (text.includes('mcp')) {
    return 'mcp';
  }
  if (text.includes('app') || text.includes('application')) {
    return 'full-stack';
  }

  // Check form data
  const data = message.content.data as {
    action?: string;
    projectId?: string;
    projectType?: string;
    formId?: string;
    formData?: any;
  };
  if (data?.projectType) {
    return data.projectType;
  }

  // Check recent memories for context
  const recentMemories = await runtime.getMemories({
    roomId: message.roomId,
    count: 10,
    unique: false,
    tableName: 'messages',
  });

  for (const memory of recentMemories) {
    const memText = memory.content.text?.toLowerCase() || '';
    if (memText.includes('plugin')) {
      return 'plugin';
    }
    if (memText.includes('agent')) {
      return 'agent';
    }
    if (memText.includes('workflow')) {
      return 'workflow';
    }
  }

  return 'plugin'; // default
}

function extractRequirements(text: string): string[] {
  const requirements: string[] = [];

  // Look for bullet points or numbered lists
  const lines = text.split('\n');
  lines.forEach((line) => {
    const trimmed = line.trim();
    if (trimmed.match(/^[-*•]/) || trimmed.match(/^\d+[.)]/)) {
      requirements.push(trimmed.replace(/^[-*•]\s*/, '').replace(/^\d+[.)]\s*/, ''));
    }
  });

  // If no list found, extract from sentences
  if (requirements.length === 0) {
    const sentences = text.match(/[^.!?]+[.!?]+/g) || [];
    sentences.forEach((sentence) => {
      if (sentence.includes('should') || sentence.includes('must') || sentence.includes('need')) {
        requirements.push(sentence.trim());
      }
    });
  }

  // If still no requirements, use the whole text
  if (requirements.length === 0 && text.length > 20) {
    requirements.push(text);
  }

  return requirements;
}

function extractAPIs(text: string): string[] {
  const apis: string[] = [];
  const apiPatterns = [
    /(\w+)\s+API/gi,
    /API[:\s]+(\w+)/gi,
    /uses?\s+(\w+)/gi,
    /integrates?\s+with\s+(\w+)/gi,
  ];

  apiPatterns.forEach((pattern) => {
    const matches = text.matchAll(pattern);
    for (const match of matches) {
      if (match[1]) {
        apis.push(match[1]);
      }
    }
  });

  return [...new Set(apis)]; // Remove duplicates
}

function extractProjectName(text: string): string | null {
  // Look for quoted names
  const quotedMatch = text.match(/["']([^"']+)["']/);
  if (quotedMatch) {
    return quotedMatch[1];
  }

  // Look for "called" or "named" patterns
  const namedMatch = text.match(/(?:called|named)\s+(\S+)/i);
  if (namedMatch) {
    return namedMatch[1];
  }

  return null;
}

function getNextStepPrompt(form: any, completedStepId: string): string {
  // Determine next step in form based on what was just completed
  const steps = form.steps || [];
  const currentStepIndex = steps.findIndex((s: any) => s.id === completedStepId);

  if (currentStepIndex >= 0 && currentStepIndex < steps.length - 1) {
    const nextStep = steps[currentStepIndex + 1];
    return `Great! Now let's ${nextStep.title}. ${nextStep.description || ''}`;
  }

  return 'Perfect! I have all the information I need. Let me start generating your code...';
}

function extractProjectData(formData: any): any {
  // Extract and structure project data from form
  const data: any = {
    projectName: formData.projectName || formData.project_name || 'unnamed-project',
    description: formData.description || '',
    requirements: [],
    apis: [],
    targetType: formData.projectType || 'plugin',
  };

  // Extract requirements
  if (formData.requirements) {
    data.requirements = formData.requirements.split('\n').filter((r: string) => r.trim());
  }

  // Extract APIs
  if (formData.apis) {
    data.apis = formData.apis
      .split(/[,\n]/)
      .map((a: string) => a.trim())
      .filter(Boolean);
  }

  // Extract test scenarios
  if (formData.test_scenarios) {
    data.testScenarios = formData.test_scenarios.split('\n').filter((s: string) => s.trim());
  }

  // GitHub settings
  data.createGithubRepo = formData.create_github === true || formData.create_github === 'true';
  data.privateRepo = formData.private_repo === true || formData.private_repo === 'true';

  // Agent-specific data
  if (formData.agent_name) {
    data.projectName = formData.agent_name;
    data.targetType = 'agent';
  }

  if (formData.personality) {
    data.requirements.push(`Agent personality: ${formData.personality}`);
  }

  if (formData.capabilities) {
    data.requirements.push(`Agent capabilities: ${formData.capabilities}`);
  }

  if (formData.plugins) {
    const plugins = formData.plugins
      .split(/[,\n]/)
      .map((p: string) => p.trim())
      .filter(Boolean);
    data.requirements.push(`Include plugins: ${plugins.join(', ')}`);
  }

  return data;
}

File: cpp_merge/classified/packages/plugin-autocoder/src/actions/generate-code.cpp
#include "generate-code.h"

std::shared_ptr<Promise<string>> extractProjectType(std::shared_ptr<IAgentRuntime> runtime, std::shared_ptr<Memory> message, std::shared_ptr<State> _state)
{
    auto text = OR((message->content->text->toLowerCase()), (string_empty));
    if (text["includes"](std:("plugin"))) {
        return std:("plugin");
    }
    if (OR((text["includes"](std:("agent"))), (text["includes"](std:("bot"))))) {
        return std:("agent");
    }
    if (text["includes"](std:("workflow"))) {
        return std:("workflow");
    }
    if (text["includes"](std:("mcp"))) {
        return std:("mcp");
    }
    if (OR((text["includes"](std:("app"))), (text["includes"](std:("application"))))) {
        return std:("full-stack");
    }
    auto data = as<object>(message->content->data);
    if (data["projectType"]) {
        return data["projectType"];
    }
    auto recentMemories = std::async([=]() { runtime->getMemories(object{
        object::pair{std:("roomId"), message->roomId}, 
        object::pair{std:("count"), 10}, 
        object::pair{std:("unique"), false}, 
        object::pair{std:("tableName"), std:("messages")}
    }); });
    for (auto& memory : recentMemories)
    {
        auto memText = OR((memory->content->text->toLowerCase()), (string_empty));
        if (memText["includes"](std:("plugin"))) {
            return std:("plugin");
        }
        if (memText["includes"](std:("agent"))) {
            return std:("agent");
        }
        if (memText["includes"](std:("workflow"))) {
            return std:("workflow");
        }
    }
    return std:("plugin");
};


array<string> extractRequirements(string text)
{
    shared requirements = array<string>();
    auto lines = text->split(std:("\
"));
    lines->forEach([=](auto line) mutable
    {
        auto trimmed = line->trim();
        if (OR((trimmed->match((new RegExp(std:("^[-*•"))))), (trimmed->match((new RegExp(std:("^\d+[.)"))))))) {
            requirements->push(trimmed->replace((new RegExp(std:("^[-*•]\s"))), string_empty)->replace((new RegExp(std:("^\d+[.)]\s"))), string_empty));
        }
    }
    );
    if (requirements->get_length() == 0) {
        auto sentences = OR((text->match((new RegExp(std:("[^.!?]+[.!?]+"))))), (array<any>()));
        sentences->forEach([=](auto sentence) mutable
        {
            if (OR((OR((sentence->includes(std:("should"))), (sentence->includes(std:("must"))))), (sentence->includes(std:("need"))))) {
                requirements->push(sentence->trim());
            }
        }
        );
    }
    if (AND((requirements->get_length() == 0), (text->get_length() > 20))) {
        requirements->push(text);
    }
    return requirements;
};


array<string> extractAPIs(string text)
{
    shared apis = array<string>();
    auto apiPatterns = array<std::shared_ptr<RegExp>>{ (new RegExp(std:("(\w+)\s+API/"))), (new RegExp(std:("API[:\s]+(\w+)/"))), (new RegExp(std:("uses?\s+(\w+)/"))), (new RegExp(std:("integrates?\s+with\s+(\w+)/"))) };
    apiPatterns->forEach([=](auto pattern) mutable
    {
        auto matches = text->matchAll(pattern);
        for (auto& match : matches)
        {
            if (const_(match)[1]) {
                apis->push(const_(match)[1]);
            }
        }
    }
    );
    return array<any>{ std::make_shared<Set>(apis) };
};


any extractProjectName(string text)
{
    auto quotedMatch = text->match((new RegExp(std:("["']([^"']+)["'"))));
    if (quotedMatch) {
        return (*const_(quotedMatch))[1];
    }
    auto namedMatch = text->match((new RegExp(std:("(?:called|named)\s+(\S+)"))));
    if (namedMatch) {
        return (*const_(namedMatch))[1];
    }
    return nullptr;
};


string getNextStepPrompt(any form, string completedStepId)
{
    auto steps = OR((form["steps"]), (array<any>()));
    auto currentStepIndex = steps["findIndex"]([=](auto s) mutable
    {
        return s["id"] == completedStepId;
    }
    );
    if (AND((currentStepIndex >= 0), (currentStepIndex < steps["length"] - 1))) {
        auto nextStep = const_(steps)[currentStepIndex + 1];
        return std:("Great! Now let's ") + nextStep["title"] + std:(". ") + (OR((nextStep["description"]), (string_empty))) + string_empty;
    }
    return std:("Perfect! I have all the information I need. Let me start generating your code...");
};


any extractProjectData(any formData)
{
    auto data = object{
        object::pair{std:("projectName"), OR((OR((formData["projectName"]), (formData["project_name"]))), (std:("unnamed-project")))}, 
        object::pair{std:("description"), OR((formData["description"]), (string_empty))}, 
        object::pair{std:("requirements"), array<any>()}, 
        object::pair{std:("apis"), array<any>()}, 
        object::pair{std:("targetType"), OR((formData["projectType"]), (std:("plugin")))}
    };
    if (formData["requirements"]) {
        data["requirements"] = formData["requirements"]["split"](std:("\
"))["filter"]([=](auto r) mutable
        {
            return r->trim();
        }
        );
    }
    if (formData["apis"]) {
        data["apis"] = formData["apis"]["split"]((new RegExp(std:("[,\n"))))["map"]([=](auto a) mutable
        {
            return a->trim();
        }
        )["filter"](Boolean);
    }
    if (formData["test_scenarios"]) {
        data["testScenarios"] = formData["test_scenarios"]["split"](std:("\
"))["filter"]([=](auto s) mutable
        {
            return s->trim();
        }
        );
    }
    data["createGithubRepo"] = OR((formData["create_github"] == true), (formData["create_github"] == std:("true")));
    data["privateRepo"] = OR((formData["private_repo"] == true), (formData["private_repo"] == std:("true")));
    if (formData["agent_name"]) {
        data["projectName"] = formData["agent_name"];
        data["targetType"] = std:("agent");
    }
    if (formData["personality"]) {
        data["requirements"]["push"](std:("Agent personality: ") + formData["personality"] + string_empty);
    }
    if (formData["capabilities"]) {
        data["requirements"]["push"](std:("Agent capabilities: ") + formData["capabilities"] + string_empty);
    }
    if (formData["plugins"]) {
        auto plugins = formData["plugins"]["split"]((new RegExp(std:("[,\n"))))["map"]([=](auto p) mutable
        {
            return p->trim();
        }
        )["filter"](Boolean);
        data["requirements"]["push"](std:("Include plugins: ") + plugins["join"](std:(", ")) + string_empty);
    }
    return data;
};


std::shared_ptr<Action> generateCodeAction = object{
    object::pair{std:("name"), std:("GENERATE_CODE")}, 
    object::pair{std:("similes"), array<string>{ std:("BUILD_CODE"), std:("CREATE_CODE"), std:("WRITE_CODE"), std:("DEVELOP_CODE"), std:("IMPLEMENT_CODE"), std:("CODE_GENERATION"), std:("GENERATE_PROJECT"), std:("BUILD_PROJECT") }}, 
    object::pair{std:("description"), std:("Generate code for ElizaOS projects using AI")}, 
    object::pair{std:("validate"), [=](auto runtime, auto message, auto _state = undefined) mutable
    {
        shared text = OR((message->content->text->toLowerCase()), (string_empty));
        auto data = as<object>(message->content->data);
        if (data["action"] == std:("GENERATE_CODE")) {
            return true;
        }
        auto genKeywords = array<string>{ std:("generate"), std:("create"), std:("build"), std:("write"), std:("implement"), std:("code") };
        auto contextKeywords = array<string>{ std:("plugin"), std:("agent"), std:("action"), std:("provider"), std:("service") };
        auto hasGenKeyword = genKeywords->some([=](auto keyword) mutable
        {
            return text["includes"](keyword);
        }
        );
        auto hasContextKeyword = contextKeywords->some([=](auto keyword) mutable
        {
            return text["includes"](keyword);
        }
        );
        return AND((hasGenKeyword), (hasContextKeyword));
    }
    }, 
    object::pair{std:("handler"), [=](auto runtime, auto message, auto state = undefined, auto _options = undefined, auto callback = undefined, auto _responses = undefined) mutable
    {
        elizaLogger->info(std:("GENERATE_CODE action triggered"));
        auto codeGenService = as<std::shared_ptr<CodeGenerationService>>(runtime->getService(std:("code-generation")));
        if (!codeGenService) {
            elizaLogger->error(std:("Code generation service not available"));
            if (callback) {
                std::async([=]() { callback(object{
                    object::pair{std:("text"), std:("I can't generate code right now - the code generation service isn't available. Please ensure all required API keys are configured.")}, 
                    object::pair{std:("error"), true}
                }); });
            }
            return object{
                object::pair{std:("success"), false}, 
                object::pair{std:("error"), std:("Code generation service not available")}
            };
        }
        auto formsService = as<any>(runtime->getService(std:("forms")));
        auto projectType = std::async([=]() { extractProjectType(runtime, message, state); });
        elizaLogger->info(std:("Project type: ") + projectType + string_empty);
        if (formsService) {
            auto activeForms = std::async([=]() { formsService["listForms"](std:("active")); });
            auto projectForm = activeForms["find"]([=](auto f) mutable
            {
                return f->id->startsWith(std:("project-"));
            }
            );
            if (AND((projectForm), (projectForm["status"] != std:("completed")))) {
                auto nextStepPrompt = getNextStepPrompt(projectForm, OR((message->content->text), (string_empty)));
                if (callback) {
                    std::async([=]() { callback(object{
                        object::pair{std:("text"), nextStepPrompt}, 
                        object::pair{std:("form"), projectForm}
                    }); });
                }
                return object{
                    object::pair{std:("success"), true}, 
                    object::pair{std:("data"), object{
                        object::pair{std:("status"), std:("collecting_requirements")}, 
                        object::pair{std:("formId"), projectForm["id"]}
                    }}
                };
            }
            auto messageData = as<object>(message->content->data);
            if (AND((messageData["formId"]), (messageData["formData"]))) {
                auto formData = messageData["formData"];
                auto projectData = extractProjectData(formData);
                if (callback) {
                    std::async([=]() { callback(object{
                        object::pair{std:("text"), std:("🚀 Starting code generation for your ") + projectData["targetType"] + std:("...\
\
This may take a few minutes. I'll:\
1. Research best practices\
2. Generate the code structure\
3. Implement all features\
4. Run quality checks")}
                    }); });
                }
                auto result = std::async([=]() { codeGenService->generateCode(object{
                    object::pair{std:("projectName"), projectData["projectName"]}, 
                    object::pair{std:("description"), projectData["description"]}, 
                    object::pair{std:("requirements"), projectData["requirements"]}, 
                    object::pair{std:("apis"), OR((projectData["apis"]), (array<any>()))}, 
                    object::pair{std:("targetType"), projectData["targetType"]}, 
                    object::pair{std:("testScenarios"), OR((projectData["testScenarios"]), (array<any>()))}
                }); });
                if (!result->success) {
                    if (callback) {
                        std::async([=]() { callback(object{
                            object::pair{std:("text"), std:("❌ Code generation failed: ") + (OR((result->errors->join(std:("\
"))), (std:("Unknown error")))) + string_empty}, 
                            object::pair{std:("error"), true}
                        }); });
                    }
                    return object{
                        object::pair{std:("success"), false}, 
                        object::pair{std:("error"), OR((result->errors->join(std:("\
"))), (std:("Code generation failed")))}
                    };
                }
                auto successMessage = std:("✅ **Code generation complete!**\
\
");
                successMessage += std:("📁 **Project:** ") + projectData["projectName"] + std:("\
");
                successMessage += std:("📋 **Type:** ") + projectData["targetType"] + std:("\
");
                if (result->files) {
                    successMessage += std:("📄 **Files generated:** ") + result->files->get_length() + std:("\
");
                }
                if (result->projectPath) {
                    successMessage += std:("\
📍 **Project location:** ") + result->projectPath + string_empty;
                }
                if (result->executionResults) {
                    successMessage += std:("\
\
**Quality Checks:**\
");
                    successMessage += string_empty + (result->executionResults->lintPass) ? std:("✅") : std:("❌") + std:(" Linting\
");
                    successMessage += string_empty + (result->executionResults->typesPass) ? std:("✅") : std:("❌") + std:(" Type checking\
");
                    successMessage += string_empty + (result->executionResults->testsPass) ? std:("✅") : std:("❌") + std:(" Tests\
");
                    successMessage += string_empty + (result->executionResults->buildPass) ? std:("✅") : std:("❌") + std:(" Build\
");
                }
                successMessage += std:("\
\
🎉 Your ") + projectData["targetType"] + std:(" is ready! Let me know if you need any modifications.");
                if (callback) {
                    std::async([=]() { callback(object{
                        object::pair{std:("text"), successMessage}, 
                        object::pair{std:("data"), object{
                            object::pair{std:("projectPath"), result->projectPath}, 
                            object::pair{std:("files"), result->files}
                        }}
                    }); });
                }
                return object{
                    object::pair{std:("success"), true}, 
                    object::pair{std:("data"), object{
                        object::pair{std:("projectPath"), result->projectPath}, 
                        object::pair{std:("filesGenerated"), OR((result->files->get_length()), (0))}, 
                        object::pair{std:("executionResults"), result->executionResults}
                    }}
                };
            }
        }
        auto requirements = extractRequirements(OR((message->content->text), (string_empty)));
        if (OR((!requirements), (requirements->get_length() == 0))) {
            if (callback) {
                std::async([=]() { callback(object{
                    object::pair{std:("text"), std:("I need more details about what you want to build. Could you describe:\
- What type of project (plugin, agent, etc.)?\
- What features or functionality it should have?\
- Any specific APIs or services it should use?")}, 
                    object::pair{std:("error"), false}
                }); });
            }
            return object{
                object::pair{std:("success"), false}, 
                object::pair{std:("error"), std:("Insufficient requirements provided")}
            };
        }
        auto projectName = OR((extractProjectName(OR((message->content->text), (string_empty)))), (std:("project-") + Date->now() + string_empty));
        if (callback) {
            std::async([=]() { callback(object{
                object::pair{std:("text"), std:("🚀 Starting code generation based on your requirements...\
\
Generating: ") + projectName + std:("\
Type: ") + projectType + string_empty}
            }); });
        }
        auto result = std::async([=]() { codeGenService->generateCode(object{
            object::pair{std:("projectName"), std:("projectName")}, 
            object::pair{std:("description"), OR((message->content->text), (string_empty))}, 
            object::pair{std:("requirements"), std:("requirements")}, 
            object::pair{std:("apis"), extractAPIs(OR((message->content->text), (string_empty)))}, 
            object::pair{std:("targetType"), as<ProjectType>(projectType)}
        }); });
        if (!result->success) {
            if (callback) {
                std::async([=]() { callback(object{
                    object::pair{std:("text"), std:("❌ Code generation failed: ") + (OR((result->errors->join(std:("\
"))), (std:("Unknown error")))) + string_empty}, 
                    object::pair{std:("error"), true}
                }); });
            }
            return object{
                object::pair{std:("success"), false}, 
                object::pair{std:("error"), OR((result->errors->join(std:("\
"))), (std:("Code generation failed")))}
            };
        }
        auto responseText = std:("✅ Code generation complete!\
\
");
        responseText += std:("📁 Project: ") + projectName + std:("\
");
        if (result->files) {
            responseText += std:("📄 Files generated: ") + result->files->get_length() + std:("\
");
        }
        if (result->projectPath) {
            responseText += std:("📍 Location: ") + result->projectPath + std:("\
");
        }
        responseText += std:("\
🎉 Your code is ready!");
        if (callback) {
            std::async([=]() { callback(object{
                object::pair{std:("text"), responseText}, 
                object::pair{std:("data"), result}
            }); });
        }
        return object{
            object::pair{std:("success"), true}, 
            object::pair{std:("data"), result}
        };
    }
    }, 
    object::pair{std:("examples"), array<array<any>>{ array<object>{ object{
        object::pair{std:("name"), std:("{{user1}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("Generate a plugin that monitors crypto prices and sends alerts")}
        }}
    }, object{
        object::pair{std:("name"), std:("{{agent}}")}, 
        object::pair{std:("content"), object{
            object::pair{std:("text"), std:("I'll generate a crypto price monitoring plugin for you. Let me create the code structure with price tracking and alert functionality.")}, 
            object::pair{std:("action"), std:("GENERATE_CODE")}
        }}
    } } }}
};

void Main(void)
{
}

MAIN
